### PR TITLE
Part1

### PR DESCRIPTION
**Pull request by Pryhym**

### CODE REVIEW

File: Character.cpp [hunk 1]
```diff
@@ -98,20 +106,44 @@ void Character::attackInternal(Character& other)
             b) your stats are boosted 10%
             c) the initial value of your stats is updated to reflect this boosted stat for the next time you defeat another character.
       */
-        assert(false);
-        std::cout << getName() << " defeated " << other.getName() << " and leveled up!" << std::endl;        
+        initialStats(hitPoints, *initialHitPoints);
+        initialStats(armor, *initialArmorLevel);
+        initialStats(attackDamage, *initialAttackDamage);
```
Comment (matkatmusic) [inline Character.cpp]: Name things what they are, name functions what they do.
`initialStats` is a thing.  it doesn't say what it does.
- [x] revise

File: Character.cpp [hunk 2]
```diff
@@ -98,20 +106,44 @@ void Character::attackInternal(Character& other)
...
+    {
+        hitPoints += amount;
+        std::cout << getName() << "'s hit point level has been boosted to " << hitPoints << std::endl;
+    }
+    
+    void Character::boostAttackDamage( double amount )
+    {
+        attackDamage += amount;
+        std::cout << getName() << "'s attack damage level has been boosted to " << attackDamage << std::endl;
+    }
```
Comment (matkatmusic) [inline Character.cpp]: None of these functions needed to be moved from the header file.
Don't change code that doesn't need to be changed.
- [x] revise

File: Dragon.cpp [hunk 3]
```diff
@@ -1,7 +1,12 @@
 #include "Dragon.h"
 #include "DragonSlayer.h"
+#include "Utility.h"
 
-//Dragon::Dragon
+Dragon::Dragon(std::string name_, int hitPoints_, int armor_) : Character(hitPoints_, armor_,80),
```
Comment (matkatmusic) [inline Dragon.cpp]: add space after `,` to make the code more readable
- [x] revise

File: DragonSlayer.cpp [hunk 4]
```diff
@@ -1,21 +1,34 @@
 #include "DragonSlayer.h"
 #include "Dragon.h"
+#include "Utility.h"
 
-//DragonSlayer::DragonSlayer
+DragonSlayer::DragonSlayer(std::string name_, int hitPoints_, int armor_) : Character(hitPoints_, armor_, 4), name(name_), attackItem(new AttackItem())
```
Comment (matkatmusic) [inline DragonSlayer.cpp]: You can initialize your attack item directly where it is declared.
You don't need to initialize it in the Member Initializer List
- [x] revise

File: DragonSlayer.cpp [hunk 5]
```diff
@@ -1,21 +1,34 @@
...
     if( auto* dragon = dynamic_cast<Dragon*>(&other) )
     {
-        assert(false);
+        
         //DragonSlayers get a 10x boost when attacking dragons, from their attack item.
         //so they should USE their attack item before attacking the dragon... 
         //note: they should only use the item if the dragon's hitpoints are > 0...
         //note: items are single-use only, so you need to reset it after use.  
         //look in the Character class for how the other item types are reset after use.
+        if (attackItem != nullptr && dragon->getHP() > 0)
```
Comment (matkatmusic) [inline DragonSlayer.cpp]: Does it makes sense for the Dragon to attack the dragon slayer if its hit points were 0?
- [x] revise

File: Item.cpp [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#include "Item.h"
+
+Item::Item(const std::string& name_, int effect ) : name(name_), boost(effect) { }
+Item:: ~Item() { }
```
Comment (matkatmusic) [inline Item.cpp]: don't add space around `::`
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: DragonSlayer.cpp (hunk 5 above)
Comment (Pryhym) [reply DragonSlayer.cpp]: this is the dragon slayer attacking the dragon

File: Character.cpp [hunk 7]
```diff
@@ -86,8 +86,16 @@ int Character::takeDamage(int damage)
     return hitPoints;
 }
 
+void updateStats(int &curStat, int &intStat)
```
Comment (matkatmusic) [inline Character.cpp]: - [x] revise, make it look like what it is:  declarations of references.

File: DragonSlayer.cpp [hunk 8]
```diff
@@ -1,21 +1,34 @@
...
+        Character::boostAttackDamage(boostAttack);
         //DragonSlayers get a 10x boost when attacking dragons, from their attack item.
         //so they should USE their attack item before attacking the dragon... 
         //note: they should only use the item if the dragon's hitpoints are > 0...
         //note: items are single-use only, so you need to reset it after use.  
         //look in the Character class for how the other item types are reset after use.
+        if ( attackItem != nullptr && dragon->getHP() < 0 )
+        {
+            attackItem->use(this);
+        }
```
Comment (matkatmusic) [inline DragonSlayer.cpp]: The attack item is what performs the boost. 
You need to reset it after it is used.
Also, if the dragon's hit points are less than zero, what is the point of using the attack item? 
- [x] revise

File: DragonSlayer.h [hunk 9]
```diff
@@ -1,8 +1,19 @@
...
+#include "AttackItem.h"
 
 struct DragonSlayer : Character
 {
+    DragonSlayer(std::string name_, int hitPoints_, int armor_);
+
+    const std::string& getName() override;
+    std::string getStats() override;
     void attack(Character& other) override;
+    AttackItem* attackItem = new AttackItem();
```
Comment (matkatmusic) [inline DragonSlayer.h]: Your'e storing heap allocated memory in a raw pointer. 
And you aren't freeing it either when the class goes out of scope. 
```
        //note: items are single-use only, so you need to reset it after use.  
        //look in the Character class for how the other item types are reset after use.
```
Take a look at how the other items are stored. 
Remember, you're only storing one, so a vector is the wrong choice.
- [x] revise

File: Character.h [hunk 10]
```diff
@@ -35,32 +35,26 @@ struct Character
...
         hitPoints += amount;
         std::cout << getName() << "'s hit point level has been boosted to " << hitPoints << std::endl;
     }
     
-    void boostAttackDamage( int amount )
+    void boostAttackDamage( double amount )
     {
         attackDamage += amount;
         std::cout << getName() << "'s attack damage level has been boosted to " << attackDamage << std::endl;
     }
```
Comment (matkatmusic) [inline Character.h]: These did not need revising. 
Don't do more work than you are asked to do.
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]
Almost there. 
But you aren't following best practices or using language features that you've previously learned correctly.

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**